Feature/blui 3667 react test doc (#921)
etn-ccis/blui-react-component-library: docs/src/markdownDocs/gettingStarted/reactTesting.mdx

Vitest offers compatibility with most of the Jest API and ecosystem libraries and transitioning to Vitest API is highly compatible with Jests API, so existing Jest knowledge can be applied.

If your team prefers Jest, it is possible to use Jest in Vite React setups. [Vite-jest](https://www.npmjs.com/package/vite-jest) integration for Vite and Jest, so this is a valid option for your unit tests but the preference for Brightlayer UI React Vite applications will be to use Vitest testing framework.

### React Vite

The Brightlayer UI [React Vite template](https://github.com/etn-ccis/blui-react-cli-templates) projects are pre-configured with [Vitest](https://v3.vitejs.dev/guide/) testing framework. When creating a new React Vite project from Brightlayer UI templates, a sample test is created in your project for you (test files are identified by the `.spec.tsx` or `.test.tsx` file extension). You can execute the tests by running the following in your terminal:

<CodeBlock code={`npm run test`} language={'sh'} />
or
<CodeBlock code={`yarn test`} language={'sh'} />

> This will run your tests and show the pass / fail output in the terminal.

Reviewer: ```suggestion
This will run your tests and show the pass/fail output in the terminal.
```

There are additional utilities that complement Vitest well. See the [Vitest Environment](https://vitest.dev/guide/environment.html#test-environment) and [features](https://vitest.dev/guide/features.html).
For testing React components in a Vite project using Vitest, you'll primarily use the following libraries:

1. [Vitest](https://v3.vitejs.dev/guide/): The testing framework itself, which is a fast alternative to Jest, designed specifically for Vite projects.
2. [React Testing Library](https://github.com/testing-library/react-testing-library#readme): A library that encourages testing your components from the user's perspective, interacting with them as a real user would.
3. [Jest-dom](https://github.com/testing-library/jest-dom#readme): A library providing custom matchers for testing DOM elements, making your assertions more readable and maintainable.

# End-to-End Testing

> End-to-end tests at a step above unit tests, both in terms of complexity and what they test. They focus primarily on the flow of the application, making sure that a user's journey through the application has the expected outcomes. These tests can generally be categorized as:

Reviewer: ```suggestion
End-to-end tests are a step above unit tests, both in terms of complexity and what they test. They focus primarily on the flow of the application, making sure that a user's journey through the application has the expected outcomes. These tests can generally be categorized as:
```

> -   **Acceptance Tests**: verify that various features and flows meet the customer expectations.

Reviewer: Same comment about `Verify`

-   **Regression Tests**: verify that existing functionality is not broken when new features are added.

> They should cover user stories that span multiple components and views, such as signing up accounts, logging in and out of the application, updating a profile, etc. These tests are not concerned with the underlying state or implementation of the application - just the end results.

Reviewer: ```suggestion
They should cover user stories that span multiple components and views, such as signing up for an account, logging in and out of the application, updating a profile, etc. These tests are not concerned with the underlying state or implementation of the application - just the end results.
```

End-to-end tests usually run in a browser against a live system (in the case of web, using a test runner that automates the browser). Automated E2E tests can take a long time to run because of the different pieces involved and the scope of the tests. You should try to optimize your testing pipeline to run tests in parallel to finish faster.

Functional tests alone will not give you enough test coverage to avoid regressions. Unit tests aim to provide code coverage depth, while functional tests provide coverage over the breadth of user test scenarios.

## How to Write

End-to-end tests rely on the ability to find elements on the screen and interact with them in an automated way. This means that you need a reliable mechanism for selecting elements. Depending on your testing framework, there are different ways to select elements, including:

> -   **CSS Selectors**: find elements by using CSS classes or rules (e.g., `.your-class-name`)

Reviewer: Same comment for `Find`. Also add period to the sentences

-   **Element Selectors**: find elements inside the elements (e.g., `#\@\@blui-drawerlayout-content > div > header > div > button`)
-   **xPath Selectors**: find selector within xPath (e.g., `//*[@id="@@blui-drawerlayout-content"]/div/header/div/button`)
-   **ID Selectors**: find attribute id within elements (e.g., `#login-button`)

> You can read more about different selector strategies on [Selenium Dev](https://www.selenium.dev/documentation/webdriver/elements/locators/) and {/* <!-- markdown-link-check-disable --> */}[software testing help](https://www.softwaretestinghelp.com/using-selenium-xpath-and-other-locators-selenium-tutorial-5/){/* <!-- markdown-link-check-enable --> */}.

> Each of these methods has their own advantages and disadvantages. A common pain point in E2E test automation is that modern Javascript platforms are constantly changing, particularly in the open source community. Many of these frameworks automatically (and dynamically) generate IDs and classes for components resulting in a constantly moving target for automation tooling. Identifying elements by these values (i.e., using IDs, CSS, or xPath locators _without_ property attributes) makes your tests brittle because they are subject to change any time a new version comes out (or even any time the page is reloaded).

Reviewer: Some grammaticals -
`open source` should be `open-source`
`Javascript` should be `JavaScript`

One way to combat this in your own applications is to use dedicated attributes for testing (e.g., `test-id` or `dev-id`). By adding this test-specific attribute to the element, identifying the correct underlying component should be safer. Because this attribute is test-specific and the application logic does not use it, it will be less likely to change throughout the product lifecycle. Combining this approach with other selectors is a common and reliable approach for element identification.

```html

> {/* Component code */}

Reviewer: <img width="838" alt="Screenshot 2024-11-12 at 3 48 53 PM" src="https://github.com/user-attachments/assets/e00888b6-b1cd-42d9-a3c2-97f98e9331f4">

This comment looks like this.
may need to html comment syntax

-   [WebDriverIO](https://webdriver.io/docs/gettingstarted.html)

### Selenium

Selenium is a popular automation testing suite which can be used to automate the desktop and mobile web browser interactions. You can write test code in any of the languages supported by Selenium.

Selenium test suite has several test frameworks available and each one can be customized to your project needs.

For more information, check out the browser stack site [selenium guide](https://www.browserstack.com/guide/selenium-framework) or the official [selenium documentation](https://github.com/SeleniumHQ/selenium).

There are also a number of [Selenium tutorials](https://www.guru99.com/selenium-tutorial.html) available online to help you get started.

### Cypress

Cypress is a popular open-source testing framework that boasts support for any modern JavaScript framework. It works well for E2E tests in React applications and is easier to use than some of the more traditional tools, such as Selenium.

For a more detailed write-up on Cypress, check out the official [Cypress Documentation](https://docs.cypress.io/guides/getting-started/installing-cypress.html#System-requirements).

### WebDriverIO

> WebdriverIO is a popular Javascript based test automation framework built on top of node.js. It is an open-source project developed for the automation testing community. WebdriverIO is extendible, compatible and feature rich.

Reviewer: `extendable` instead of `extendible`

For more information, check out the official [Webdriver.io documentation](https://webdriver.io/) and their [getting started guide](https://webdriver.io/docs/gettingstarted.html).

### Vitest Comparisons

> Read more about testing frameworks that works well with [Vitest](https://vitest.dev/guide/comparisons.html#comparisons-with-other-test-runners)

Reviewer: Period in the last
